django/db/models/query.py

```diff
@@ -888,11 +888,39 @@ def bulk_update(self, objs, fields, batch_size=None):
             obj._prepare_related_fields_for_save(
                 operation_name="bulk_update", fields=fields
             )
-        # PK is used twice in the resulting update query, once in the filter
-        # and once in the WHEN. Each field will also have one CAST.
         self._for_write = True
         connection = connections[self.db]
-        max_batch_size = connection.ops.bulk_batch_size(["pk", "pk"] + fields, objs)
+        pk_fields = self.model._meta.pk_fields
+        # The query that will be generated here is of the form:
+        # ```
+        # UPDATE
+        # SET f0 = (
+        #   CASE WHEN (pk0 = pkv0 ... AND ... pkn = pkn_vo) THEN f0_v0
+        #   ...
+        #   CASE WHEN (pk0 = pk0_vm ... AND ... pkn = pkn_vm) THEN f0_v0
+        # )
+        # ...
+        # SET fk = (
+        #   CASE WHEN (pk0 = pkv0 ... AND ... pkn = pkn_vo) THEN fk_v0
+        #   ...
+        #   CASE WHEN (pk0 = pk0_vm ... AND ... pkn = pkn_vm) THEN fk_v0
+        # )
+        # WHERE
+        #   (pk0 = pk0_v0 ... AND ... pkn = pkn_v0) OR
+        #   ...
+        #   (pk0 = pk0_vm ... AND ... pkn = pkn_vm)
+        # ```
+        # For k = len(fields), m = len(objs), and n = len(pk_fields).
+        # So the number of required placeholders will be:
+        # 1. len(objs) * len(pk_fields) * len(fields) for each field of each object the
+        #    primary key fields must be specified for the CASE / WHEN criteria.
+        # 2. len(objs) * len(fields) each field needs to be assigned a different value
+        #    for each object.
+        # 3. len(objs) * len(pk_fields) for the WHERE clause.
+        # So the placeholders in the query are:
+        # ((len(pk_fields) * len(fields)) + len(fields) + len(pk_fields)) * len(objs)
+        fields_in_query = ((pk_fields * len(fields)) + fields + pk_fields) * len(objs)
```

Comment:
I really don't think I should be multiplying by the length of objects, however, when I remove this I get:
```
django.db.utils.OperationalError: Expression tree is too large (maximum depth 1000)
```
Which I think is dependent on `SQLITE_MAX_EXPR_DEPTH` which defaults to 1000

Reply:
My current theory is that maybe the Oracle `max_query_param` value should be smaller due to the 8000 byte query size limit :thinking: 
https://docs.oracle.com/cd/E19851-01/819-7557/gfqsr/index.html